### PR TITLE
Make Generator aware of AutoScheduler

### PR DESCRIPTION
**Pull request by steven-johnson**
This revises the way AutoSchedule is integrated into Generator, with the intent of making it easier to use and harder to misuse; these changes will also pave the way to using Stubs with the auto-scheduler in a seamless way (in a subsequent PR).

Relevant changes:
-- Just as all Generators now have an implicit GeneratorParam for Target, they now have one for 'auto_schedule' and 'machine_params'. Also like 'target', these GeneratorParams are always propagated into sub-generators automatically.
-- Generators should no longer call auto_schedule_outputs() themselves (and in fact, this call is removed); Halide will call this for you after the toplevel Generator has finished.
-- You can still see the auto-schedule output in the debug(3) output, but the now-preferred way to get the output is by adding 'schedule' to -e, which saves it to a file named ${GENNAME}.schedule (basically, analogous to saving .stmt, etc).
-- Note that the auto-schedule output now includes the Target and MachineParams for which it was produced as comments.

I updated tutorial lesson 21 as I thought appropriate but it definitely needs a careful eye, esp. regarding MachineParams.

### CODE REVIEW

**Comment by steven-johnson** [comment]
I added a new app ("linear_blur") which demonstrates usage of auto-schedule with stubs:
-- Add trivial srgb-to-linear, linear-to-srgb, and blur3x3 generators
-- Add a "linear blur" generator that fuses all three of these together via auto-schedule
-- Add a demo app that uses concrete impls of both linear-blur and the basic blur3x3, with buildfiles in CMake and Bazel

PTAL

**Comment by steven-johnson** [comment]
I'm honestly a little unsure about adding more "magic" predefined GeneratorParams. I initially toyed with having auto_schedule and machine_params be specified by an additional flag (e.g.` --auto_schedule --machine_params=1,2,3`) to the Generator binary, which would then be mirrored into new arguments to the CMake/Bazel rules (e.g. `auto_schedule=true`), but after putting that in place, it felt like it was basically recapitulating code from GeneratorParams that was already handling this sort of thing.

Another alternative (which I didn't try) would be to skip having the GeneratorParams built-in, but allowing them to optionally be present, and having the generation code dynamically inquire of their presence ("if auto_schedule is present and true"...). That seems even weirder to me.

File: bazel/halide.bzl [hunk 1]
```diff
@@ -297,6 +297,7 @@ _output_extensions = {
     "assembly": ("s.txt", True),
     "bitcode": ("bc", True),
     "stmt": ("stmt", True),
+    "schedule": ("schedule", True),
```
Comment (zvookin) [inline bazel/halide.bzl]: Should this have a docs update to README_bazel.md?

**Comment by steven-johnson** [comment]
I like the idea of moving linear_blur to a separate PR. Let me take a look at that.

**Comment by steven-johnson** [comment]
(This PR is being split into smaller pieces.)

File: Makefile [hunk 2]
```diff
@@ -1349,11 +1349,19 @@ $(BIN_DIR)/tutorial_lesson_21_auto_scheduler_generate: $(ROOT_DIR)/tutorial/less
 	$(CXX) $(TUTORIAL_CXX_FLAGS) $(IMAGE_IO_CXX_FLAGS) $(OPTIMIZE) $< $(BUILD_DIR)/GenGen.o \
 	-I$(INCLUDE_DIR) $(TEST_LD_FLAGS) $(IMAGE_IO_LIBS) -o $@
 
+# The values in MachineParams are:
+# - the maximum level of parallelism available, 
+# - the size of the last-level cache (in KB), 
+# - the ratio between the cost of a miss at the last level cache and the cost
+#   of arithmetic on the target architecture
+# ...in that order.
```
Comment (steven-johnson) [inline Makefile]: Added documentation per @zalman request, but looking again, the existing comment (copied from elsewhere) looks wrong: given the default, surely the cache size is in bytes, not kilobytes? (cc @psuriana)

**Comment by steven-johnson** [comment]
Finally ready for review, I think; PTAL

**Comment by zvookin** [comment]
We need to add documentation for the machine params and auto schedule generator params. Not sure where, but presumably in the same place target is documented. Otherwise, LGTM.

**Comment by steven-johnson** [comment]
> presumably in the same place target is documented 

which is to say, largely undocumented :-(

Working on it.

**Pull request merged**